File: paper/src/main/java/com/gitlab/aecsocket/minecommons/paper/effect/ParticleEffect.java
package com.gitlab.aecsocket.minecommons.paper.effect;

import com.gitlab.aecsocket.minecommons.core.vector.cartesian.Vector3;
import org.checkerframework.checker.nullness.qual.Nullable;

/**
 * A Paper implementation of a particle effect.
 */
public record ParticleEffect(
        org.bukkit.Particle name,
        int count,
        Vector3 size,
        double speed,
        @Nullable Object data
) implements com.gitlab.aecsocket.minecommons.core.effect.ParticleEffect {
    @Override
    public String toString() {
        return name.name() + " x" + count + " " + size + " @ " + speed +
                (data == null ? "" : "[" + data + "]");
    }
}

File: core/src/main/kotlin/com/gitlab/aecsocket/alexandria/core/effect/ParticleEffect.kt
package com.gitlab.aecsocket.alexandria.core.effect

import com.gitlab.aecsocket.alexandria.core.physics.Vector3
import net.kyori.adventure.key.Key

data class ParticleEffect(
    val particle: Key,
    val count: Double = 0.0,
    val size: Vector3 = Vector3.Zero,
    val speed: Double = 0.0,
    val data: Any? = null
)
